### ISSUE TITLE
error at MiniPBCoder.cpp:275::decodeOneMap

### ISSUE THREAD

**Issue by samlu** [Tencent/MMKV#482]
### The language of MMKV
Java

### The version of MMKV
1.2.0

### The platform of MMKV
Android

### The installation of MMKV
aar

### What's the issue?
Got an error at MiniPBCoder.cpp:275::decodeOneMap

1. run app with mmkv 1.1.2 for creating a mmkv file (no encryption with MULTI_PROCESS_MODE)
2. run app with mmkv 1.2.0 
3. close app and run it again will get the following errors
-> mmkv data is reset to empty (caused by this error?)
4. close and run the app again seems okay but the old data is missed

```
07-01 15:22:04.431 I/MMKV: <MMKV_IO.cpp:271::checkLoadData> [70d28af2d18d9ced41dcadd0432c15e5] oldSeq 1104, newSeq 1105
07-01 15:22:04.431 I/MMKV: <MMKV.cpp:241::clearMemoryCache> clearMemoryCache [70d28af2d18d9ced41dcadd0432c15e5]
07-01 15:22:04.432 I/MMKV: <MMKV_IO.cpp:79::loadFromFile> loading [70d28af2d18d9ced41dcadd0432c15e5] with 25928 actual size, file size 32768, InterProcess 1, meta info version:3
07-01 15:22:04.432 I/MMKV: <MMKV_IO.cpp:84::loadFromFile> loading [70d28af2d18d9ced41dcadd0432c15e5] with crc 2176800716 sequence 1105 version 3
07-01 15:22:04.433 E/MMKV: <MiniPBCoder.cpp:275::decodeOneMap> reach end, m_position: 25928, m_size: 25928
07-01 15:22:04.434 I/MMKV: <MMKV_IO.cpp:124::loadFromFile> loaded [70d28af2d18d9ced41dcadd0432c15e5] with 0 values

07-01 15:22:55.781 I/MMKV: <MMKV_Android.cpp:145::mmkvWithID> prepare to load @@main.log (id 18a89c1b407bdac635909457a579a1cc) from relativePath /data/user/0/xxx
07-01 15:22:55.795 I/MMKV: <MMKV_IO.cpp:79::loadFromFile> loading [18a89c1b407bdac635909457a579a1cc] with 34629 actual size, file size 65536, InterProcess 0, meta info version:3
07-01 15:22:55.795 I/MMKV: <MMKV_IO.cpp:84::loadFromFile> loading [18a89c1b407bdac635909457a579a1cc] with crc 2815101864 sequence 153 version 3
07-01 15:22:55.802 E/MMKV: <MiniPBCoder.cpp:275::decodeOneMap> reach end, m_position: 34629, m_size: 34629
07-01 15:22:55.802 I/MMKV: <MMKV_IO.cpp:124::loadFromFile> loaded [18a89c1b407bdac635909457a579a1cc] with 0 values
```

**Comment by lingol** [comment]
Which arch do you running? armv7 or arm64?

**Comment by lingol** [comment]
I can't reproduce this problem. Could you send me a demo project?

**Comment by samlu** [comment]
I can see the same error on either armabi_v7a (runs Android 6) or arm64_v8a (runs Android 9).

**Comment by lingol** [comment]
I can't reproduce this problem. Could you send me a demo project? guoling@tencent.com

**Comment by samlu** [comment]
Okay, I will try to create it. Stay tuned!

**Comment by samlu** [comment]
Sorry, I cannot reproduce the problem by using a test project either. I am sure that I can see the error for my real projects and on different devices so something must have gone wrong.

Please keep this ticket for a while to see if other developers have the same issue. Meanwhile, I will keep trying to see if I can reproduce this error.

**Comment by lingol** [comment]
There's no need to be sorry. By the way, if you can reproduce this bug in your app, here's a way that can help me debug it:
> 1. run app with mmkv 1.1.2 for creating a mmkv file (no encryption with MULTI_PROCESS_MODE)
> 2. run app with mmkv 1.2.0
> 3. close app and run it again will get the following errors
> -> mmkv data is reset to empty (caused by this error?)
>4.  close and run the app again seems okay but the old data is missed

Just stop after step 2, export that file, and make a **backup** of the file. 
Then load that file in a demo project with v1.2.0, check if that error log happens again or not.
If it does happen in the demo project, upload the backup here.

**Comment by lingol** [comment]
Note: you should export the relative CRC file as well.

**Comment by samlu** [comment]
Great! Thank you or the tip. Do you have any idea what the possible cause is?

**Comment by lingol** [comment]
> Great! Thank you or the tip. Do you have any idea what the possible cause is?

No idea until I get the corrupted file and the logs on them. Remember to upload the **complete logs** of MMKV from steps 1 & 2.

**Comment by samlu** [comment]
I now can always reproduce this error.

A. test_mmkv
   . demo project
B. test_mmkv/mmkv_files/mmkv_112/41b1db9db39953c0cef4fdebf3e1abb1
   . mmkv that created by my real app with mmkv 1.1.2
C: test_mmkv/mmkv_files/mmkv_120/41b1db9db39953c0cef4fdebf3e1abb1
   . I rebuild my real app with mmkv 1.2.0
   . open my real app, this app will open the mmkv file (B), add and delete some key-values
     and then updat to this mmkv file (C) 
   . I.e. this mmkv file is modified by my real app with mmkv 1.2.0 based on the file (B)
   
How to reproduce it
1. import test_mmkv project and run it on your device (I am using an Android 6 device)
2. exit test_mmkv and kill the process
3. copy test_mmkv/mmkv_files/mmkv_120/41b1db9db39953c0cef4fdebf3e1abb1 and crc files
   to /data/data/com.a0soft.lab.test_mmkv/files/mmkv of your Android devce
   /data/data/com.a0soft.lab.test_mmkv/files/mmkv $ ls
   41b1db9db39953c0cef4fdebf3e1abb1
   41b1db9db39953c0cef4fdebf3e1abb1.crc
4. open test_mmkv app, tap the "Retrieve KVs" button, then you should be able to see the following error
   I/MMKV: <native-bridge.cpp:108::JNI_OnLoad> current API level = 23
   I/MMKV: <MMKV.cpp:138::initialize> version v1.2.0 page size:4096
   I/MMKV: <MMKV.cpp:155::initializeMMKV> root dir: /data/user/0/com.a0soft.lab.test_mmkv/files/mmkv
   D/test_mmkv: mmkv root: /data/user/0/com.a0soft.lab.test_mmkv/files/mmkv
   I/MMKV: <MMKV_IO.cpp:79::loadFromFile> loading [41b1db9db39953c0cef4fdebf3e1abb1] with 18216 actual size, file size 32768, InterProcess 1, meta info version:3
   I/MMKV: <MMKV_IO.cpp:84::loadFromFile> loading [41b1db9db39953c0cef4fdebf3e1abb1] with crc 1216713608 sequence 412 version 3
   E/MMKV: <MiniPBCoder.cpp:275::decodeOneMap> reach end, m_position: 18216, m_size: 18216
   I/MMKV: <MMKV_IO.cpp:124::loadFromFile> loaded [41b1db9db39953c0cef4fdebf3e1abb1] with 0 values
   D/test_mmkv: total 0 keys
   D/test_mmkv: finished to check all key/values

**Comment by lingol** [comment]
And the logs, please.
>Remember to upload the complete logs of MMKV from steps 1 & 2.

**Comment by samlu** [comment]
Logcat log? If not, let me know how to get the log.

**Comment by lingol** [comment]
MMKV prints log to Logcat by default. You can filter with keyword MMKV.

**Comment by samlu** [comment]
No log of step 1 as the mmkv file was created by my real app with mmkv a few months ago. 

Here is the log of step 2, i.e. the step C (generate test_mmkv/mmkv_files/mmkv_120). 
No errors during this step.
07-03 12:19:25.528 I/MMKV: <native-bridge.cpp:108::JNI_OnLoad> current API level = 23
07-03 12:19:25.528 I/MMKV: <MMKV.cpp:138::initialize> version v1.2.0 page size:4096
07-03 12:19:25.528 I/MMKV: <MMKV.cpp:155::initializeMMKV> root dir: /data/user/0/xxx
07-03 12:19:32.156 I/MMKV: <MMKV_Android.cpp:145::mmkvWithID> prepare to load @@main.log (id 41b1db9db39953c0cef4fdebf3e1abb1) from relativePath /data/user/0/xxx
07-03 12:19:32.157 I/MMKV: <MMKV_IO.cpp:79::loadFromFile> loading [41b1db9db39953c0cef4fdebf3e1abb1] with 18524 actual size, file size 32768, InterProcess 0, meta info version:3
07-03 12:19:32.157 I/MMKV: <MMKV_IO.cpp:84::loadFromFile> loading [41b1db9db39953c0cef4fdebf3e1abb1] with crc 3314192607 sequence 411 version 3
07-03 12:19:32.160 I/MMKV: <MMKV_IO.cpp:124::loadFromFile> loaded [41b1db9db39953c0cef4fdebf3e1abb1] with 261 values

**Comment by lingol** [comment]
That's it? No logs after that? When you add and delete some key-values, there are no logs?

**Comment by lingol** [comment]
I've looked into your files. There's nothing wrong with (B), the problem must be with the step 2 that add & delete some keys.
Now I need those exact operations on step 2. Here's how you can do it:
On each code that adds or deletes from the MMKV file, you print the key & value to logcat as well. Filter them out & upload here.

**Comment by lingol** [comment]
Once I have those operation logs, I can reproduce the problem locally.

**Comment by lingol** [comment]
A second look into your file, using Hex tools, I think I found the problem. Will try to reproduce it soon.
So no need to record your operations for now.

**Comment by samlu** [comment]
I am using MMKVLogLevel.LevelDebug and that's all logs I've found at step 2. 
As I mentioned, I can see the same errors for my apps so it should not relate to what the app does at step 2.

Glad to know that you have found the cause and look forward to seeing the solution. 
Thanks in advance for this excellent library.

**Comment by lingol** [comment]
Fixed in the `dev` branch.

**Comment by lingol** [comment]
You can check out the [test case](https://github.com/Tencent/MMKV/blob/dev/Android/MMKV/mmkvdemo/src/main/java/com/tencent/mmkvdemo/MainActivity.java#L449) for a reproducible demo.

**Comment by lingol** [comment]
Fixed in [v1.2.1](https://github.com/Tencent/MMKV/releases/tag/v1.2.1).

**Comment by samlu** [comment]
Awesome! I can confirm that v1.2.1 fixes this error.
